arbos/l2pricing/l2pricing.go

```diff
@@ -208,46 +160,102 @@ func (ps *L2PricingState) setConstraintsFromLegacy() error {
 		return err
 	}
 	backlog := arbmath.SaturatingUSub(oldBacklog, arbmath.SaturatingUMul(backlogTolerance, target))
-	return ps.AddConstraint(target, adjustmentWindow, backlog)
+	return ps.AddGasConstraint(target, adjustmentWindow, backlog)
 }
 
-func (ps *L2PricingState) AddConstraint(target uint64, adjustmentWindow uint64, backlog uint64) error {
-	subStorage, err := ps.constraints.Push()
+func (ps *L2PricingState) AddGasConstraint(target uint64, adjustmentWindow uint64, backlog uint64) error {
+	subStorage, err := ps.gasConstraints.Push()
 	if err != nil {
 		return fmt.Errorf("failed to push constraint: %w", err)
 	}
-	constraint := OpenGasConstraint(subStorage)
-	if err := constraint.target.Set(target); err != nil {
+	constraint := constraints.OpenGasConstraint(subStorage)
+	if err := constraint.SetTarget(target); err != nil {
 		return fmt.Errorf("failed to set target: %w", err)
 	}
-	if err := constraint.adjustmentWindow.Set(adjustmentWindow); err != nil {
+	if err := constraint.SetAdjustmentWindow(adjustmentWindow); err != nil {
 		return fmt.Errorf("failed to set adjustment window: %w", err)
 	}
-	if err := constraint.backlog.Set(backlog); err != nil {
+	if err := constraint.SetBacklog(backlog); err != nil {
 		return fmt.Errorf("failed to set backlog: %w", err)
 	}
 	return nil
 }
 
-func (ps *L2PricingState) ConstraintsLength() (uint64, error) {
-	return ps.constraints.Length()
+func (ps *L2PricingState) GasConstraintsLength() (uint64, error) {
+	return ps.gasConstraints.Length()
+}
+
+func (ps *L2PricingState) OpenGasConstraintAt(i uint64) *constraints.GasConstraint {
+	return constraints.OpenGasConstraint(ps.gasConstraints.At(i))
+}
+
+func (ps *L2PricingState) ClearGasConstraints() error {
+	length, err := ps.GasConstraintsLength()
+	if err != nil {
+		return err
+	}
+	for range length {
+		subStorage, err := ps.gasConstraints.Pop()
+		if err != nil {
+			return err
+		}
+		constraint := constraints.OpenGasConstraint(subStorage)
+		if err := constraint.Clear(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (ps *L2PricingState) MultiGasConstraintsLength() (uint64, error) {
+	return ps.multigasConstraints.Length()
 }
 
-func (ps *L2PricingState) OpenConstraintAt(i uint64) *GasConstraint {
-	return OpenGasConstraint(ps.constraints.At(i))
+func (ps *L2PricingState) OpenMultiGasConstraintAt(i uint64) *constraints.MultiGasConstraint {
+	return constraints.OpenMultiGasConstraint(ps.multigasConstraints.At(i))
+}
+
+func (ps *L2PricingState) AddMultiGasConstraint(
+	target uint64,
+	adjustmentWindow uint64,
+	backlog uint64,
+	resourceWeights map[uint8]uint64,
```

Comment:
I'm growing more confident that we should probably use arbmath.Bips for the weights, and the sum of the weights should be 10,000 (100%).